minetest/minetest
# Add hypertext support to tooltips

### Summary

Adds fancy hypertext support to tooltips to make them look pretty.

![supertip](https://github.com/user-attachments/assets/d92171a5-3b72-4993-98bf-546877f6ba13)

### Background of this PR

I was asked by @kilbith to submit this PR, who did the work for the first commit. @kilbith is unable to do it alone because of a ban. I have looked at the new feature and tested it myself, checked the code and it looks OK. I believe it will be like a very good addition to Minetest, but it needs work. Please read the code carefully yourself before running it.

I do not know if our policies allow a PR to be made like this. If you dislike what I try to do here, I feel very sorry. Just close this PR and it's over. But if this is OK:

We have explicit, written permission to use this work under LGPLv2.1+. Quote from e-mail conversation with @kilbith:

> 1. Permission is expressly granted by my organization to license this code
> under the GNU LGPL v2.1+ terms.

This e-mail has been cc'ed to @celeron55 as well.

~~**IMPORTANT: DO NOT SQUASH!** The first commit MUST be left unaltered, which contains the requested author identity. I was given the patch only under this condition, to which I agreed. The agreement allows to *add* new commits to build on the initial commit tho. So bugfixes and changes all have to go in NEW commits.~~ (kilbith has retracted the no-squash requirement, so it no longer applies)

### Feature description

This feature will add hypertext support for tooltips by introducing a new `hypertip` tooltip element. This element works like `tooltip` but you can write hypertext (like for the `hypertext` element).

There are two modes: dynamic and static. Dynamic supertips will show up next to the cursor, just like normal ones. Static supertips appear at a fixed formspec coordinate.

Also, hypertext tooltips can be *styled*.

See the `lua_api.md` for details.

### How to test

There is an example mod  with an example formspec under <https://codeberg.org/Wuzzy/temporary_luanti_hypertip_test_mod> that you can install and try (see instructions over there). This URL is *temporary* and will be deleted once this PR is merged.

This PR itself also extends the `/test_formspec` command in DevTest by adding a "Tooltip" tab. This new tab demonstrates a few minimalistic normal tooltips and hypertext tooltips.

### Demo

Here is a video demonstration:

<https://www.youtube.com/watch?v=RAASW_Rj8eQ>

### Feedback

Feedback of the concept is greatly appreciated. Especially please comment on the syntax and design choices.

Also report any bugs I didn't find myself before.
